parse/src/main/java/com/parse/ParsePlugins.java

```diff
@@ -169,27 +176,170 @@ InstallationId installationId() {
         synchronized (lock) {
             if (installationId == null) {
                 installationId =
-                        new InstallationId(new File(getFilesDir(), INSTALLATION_ID_LOCATION));
+                    new InstallationId(new File(getFilesDir(), INSTALLATION_ID_LOCATION));
             }
             return installationId;
         }
     }
 
     File getCacheDir() {
         synchronized (lock) {
-            if (cacheDir == null) {
-                cacheDir = new File(applicationContext.getCacheDir(), "com.parse");
-            }
-            return createFileDir(cacheDir);
+            return _getCacheDir();
         }
     }
 
     File getFilesDir() {
         synchronized (lock) {
-            if (filesDir == null) {
-                filesDir = new File(applicationContext.getFilesDir(), "com.parse");
+            return _getFilesDir();
+        }
+    }
+
+    private File _getFilesDir() {
+        if (filesDir == null) {
+            filesDir = new File(applicationContext.getFilesDir(), "com.parse");
+        }
+        return createFileDir(filesDir);
+    }
+
+    private File _getCacheDir() {
+        if (cacheDir == null) {
+            cacheDir = new File(applicationContext.getCacheDir(), "com.parse");
+        }
+        return createFileDir(cacheDir);
+    }
+
+    /*Migration [START]*/
+
+    /**
+     * Get cache, file object for Parse.
+     *
+     * @param useMigration to migrate old data to new locations before returning cache {@link File} object.
+     */
+    File getCacheDir(Boolean useMigration) {
+        synchronized (lock) {
+            if (useMigration) {
+                runSilentMigration();
+            }
+            return _getCacheDir();
+        }
+    }
+
+    /**
+     * Get files, file object for Parse.
+     *
+     * @param useMigration to migrate old data to new locations before returning cache {@link File} object.
+     */
+    File getFilesDir(Boolean useMigration) {
+        synchronized (lock) {
+            if (useMigration) {
+                runSilentMigration();
             }
-            return createFileDir(filesDir);
+            return _getFilesDir();
         }
     }
+
+    /**
+     * Start old data migrations to new respective locations ({@link #getFilesDir()}, {@link #getCacheDir()}).
+     */
+    public void runSilentMigration() {
+        ArrayList<File> filesToBeMigrated = new ArrayList<>();
+        ParseFileUtils.getAllNestedFiles(
+            getOldParseDir().getAbsolutePath(),
+            filesToBeMigrated
+        );
+        runParseDirFilesMigrationToCacheDir(filesToBeMigrated);
+    }
+
+    public static String replaceLast(String text, String regex, String replacement) {
+        return text.replaceFirst("(?s)" + regex + "(?!.*?" + regex + ")", replacement);
```

Comment:
Why is public and why is called replace last, but we have text replace first?

Reply:
Code refactored and reason is `Before returning the path, replace file name from the last, eg. dir name & file name could be same, as we want to get only dir name.`